topdown/http.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+// DefaultHTTPRequestTimeoutSec Default timeout value for http client
+const DefaultHTTPRequestTimeoutSec = 5
+
+var allowedKeysMap map[string]bool
+var client *http.Client
+
+func builtinHTTPReq(bctx BuiltinContext, a ast.Value) (ast.Value, error) {
+	switch a.(type) {
+	case ast.Object:
+		obj, ok := a.(ast.Object)
+		if !ok {
+			return nil, fmt.Errorf("type assertion error")
+		}
+
+		return evaluateHTTPRequest(obj, bctx)
+	default:
+		return nil, builtins.NewOperandTypeErr(1, a, ast.ObjectTypeName)
+	}
+
+}
+
+func init() {
+	createHTTPClient()
+	createAllowedKeysMap()
+	RegisterFunctionalBuiltin1WithCtxt(ast.HTTPReqBuiltin.Name, builtinHTTPReq)
+}
+
+// createHTTPClient creates a HTTP client
+func createHTTPClient() {
+	var timeout int
+
+	timeoutDuration := os.Getenv("HTTP_REQUEST_TIMEOUT_SEC")
+
+	if timeoutDuration == "" {
+		timeout = DefaultHTTPRequestTimeoutSec
+	} else {
+		timeout, _ = strconv.Atoi(timeoutDuration)
+	}
+
+	client = &http.Client{
+		Timeout: time.Duration(timeout) * time.Second,
+	}
+}
+
+// createAllowedKeysMap creates a map of the keys allowed by the builtin
+func createAllowedKeysMap() {
+	allowedKeysMap = make(map[string]bool)
+	allowedKeysMap["method"] = true
+	allowedKeysMap["url"] = true
+	allowedKeysMap["body"] = true
+}
+
+// evaluateHTTPRequest executes the HTTP request and processes the response
+func evaluateHTTPRequest(obj ast.Object, bctx BuiltinContext) (ast.Value, error) {
+	var url string
+	var method string
+	var body *strings.Reader
+
+	for _, val := range obj.Keys() {
```

Comment:
This validation can be refactored to avoid all of the string manipulation. Here's an example:

Define the allowed and required keys globally:

```
var allowedKeys = ast.NewSet(ast.StringTerm("method"), ast.StringTerm("url"), ast.StringTerm("body"))
var requiredKeys = ast.NewSet(ast.StringTerm("method"), ast.StringTerm("url"))
```

Then validation becomes:

```
requestKeys := ast.NewSet(obj.Keys()...)
invalidKeys := requestKeys.Diff(allowedKeys)
missingKeys := requiredKeys.DIff(requestKeys)
```

Then you can check if invalidKeys or missingKeys are non-empty and just return an error message containing them.

Reply:
Done